Adding management of .iml files, removing .iml files from tree.

gspencergoog opened the pull request:
This PR creates a new flutter command 'ide-config' that will create and/or update existing .iml files and some files under the .idea directory.

It also:
 * Adds *.iml to the .gitignore
 * Removes existing .iml files from the repo, and moves them to the packages/flutter_tools/templates/intellij directory.
 * Adds a flag to ide-config ('--update-templates') that will take any new .iml files in the flutter tree and add them to the existing templates.
    - If --overwrite is also specified, then all existing templates will also be overwritten with the contents from the flutter tree, and any that have been deleted from the flutter tree will also be removed from the templates.
 * Added new run configurations for all existing app targets that will now also be automatically added to IntelliJ.
 * Setting up the environment also includes setting the coding style guidelines and the git VCS.

Note that after this PR lands, Flutter developers will need to run it once to re-create the .iml files and configuration files that have been removed.

After this PR lands, .iml files will no longer appear in the untracked files section for git.

Most of the PR is just moving things around or deleting things.  Main code changes are in:
 * `packages/flutter_tools/lib/src/commands/ide_config.dart`
 * `packages/flutter_tools/lib/src/template.dart`
 * `packages/flutter_tools/test/commands/ide_config_test.dart`

gspencergoog committed:
Creates a new flutter command 'ide-config' and removes *.iml and .idea from tree.

Creates a new flutter command 'ide-config' that will create and/or update
existing .iml files and some files under the .idea directory, as well as
removing existing *.iml files and the .idea directory.

It also:

 * Adds *.iml to the .gitignore
 * Removes existing .iml files from the repo, and moves them to the
   packages/flutter_tools/templates/intellij directory.
 * Adds a flag to ide-config ('--update-templates') that will take any new .iml
   files in the flutter tree and add them to the existing templates.
     - If --overwrite is also specified, then all existing templates will also
       be overwritten with the contents from the flutter tree, and any that have
       been deleted from the flutter tree will also be removed from the
       templates.
 * Added new run configurations for all existing app targets that will now also
   be automatically added to IntelliJ.
 * Setting up the environment also includes setting the coding style guidelines
   and the git VCS.
 * Note that after this PR lands, Flutter developers will need to run it once to
   re-create the .iml files and configuration files that have been removed.

After this PR lands, .iml files will no longer appear in the untracked files
section for git.

gspencergoog committed:
Removing IntelliJ config files from repo.

gspencergoog committed:
Command created.

gspencergoog committed:
Adding update-templates

a14n commented on packages/flutter_tools/lib/src/commands/ide_config.dart:
```diff
@@ -0,0 +1,221 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
```
nit: 2017

gspencergoog replied:
Thanks, done.

a14n commented on packages/flutter_tools/lib/src/commands/ide_config.dart:
```diff
@@ -0,0 +1,221 @@
+    while (path.isNotEmpty && fs.path.dirname(path) != path) {
+      if (fs.path.basename(path) == dir) {
+        return true;
+      }
```
nit: avoidable braces

>Avoid braces around one-line code blocks (e.g. in if statements) unless there's a chain (with else) and one of the elements in the chain has a multi-line block.

This nit applies at several places in the PR.

gspencergoog replied:
Actually, we recently relaxed this, the style guide only says that single-line bodies that wrap to more than one line need to have braces: the "avoid" section has been removed.  The Dart style guide is explicitly the other way, with "DO use curly braces for all flow control structures."

a14n replied:
Ok, thanks for the info.

I suppose [this linter issue](https://github.com/dart-lang/linter/issues/142) should be updated to remove what I quoted.

a14n commented on packages/flutter_tools/test/commands/ide_config_test.dart:
```diff
@@ -0,0 +1,314 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
```
nit: 2017

gspencergoog replied:
Done.

a14n commented on packages/flutter_tools/test/commands/ide_config_test.dart:
```diff
@@ -0,0 +1,314 @@
+          contents[path] = 'dir';
+        } else if (fs.isFileSync(absPath)) {
+          contents[path] = fs.file(absPath).readAsStringSync();
+        } else {}
```
unnecessary else

gspencergoog replied:
Removed.

gspencergoog committed:
Review changes

devoncarew commented:
awesome! taking a look now -

gspencergoog commented:
After thinking about it overnight, I was thinking about two things:
1) I wonder if putting it in the templates directory is a bad idea: I think I can probably create an "intellij" project with `flutter create` now, which is not what I intended.  I might move it to "ide_templates" or something.
2) I wonder if I should leave the .iml files in the examples directories intact?  It might be non-intuitive to a newbie that no .iml files are in there, and if they just copy the examples to another directory then they'll have to set that up.  On the other hand, newbies really should be using `flutter create` to create new projects.

devoncarew commented:
Yeah, without looking through the full PR, my thoughts are:
- we should keep the iml and metadata files in examples/flutter_gallery - that's a great pedagogical example, and we should keep it easy for users to open
- creating the intellij metadata for the repo is something that will only be done by flutter repo developers; we should hide it somewhat from the main `flutter` command. So I think that's either making `flutter 'ide-config` a hidden command (like `flutter update-packages`), or making it part of update-packages itself - something like `flutter update-packages --gen-intellij` or similar. I like the second one, as it keeps the UI surface area of the flutter tool smaller

gspencergoog commented:
OK, I'll make some of those changes and you can take another look.

- I think I'll just make it a hidden command: there are times when I'd like to just reset my IntelliJ setup without waiting for update-packages.
- I'll move the templates for the command into another directory so that it isn't picked up by `flutter create`.
- I'll restore the .iml and .idea files in flutter_gallery.

gspencergoog committed:
Moving templates to iml_templates, fixing some bugs.

gspencergoog commented:
OK, so after talking to some folks to gather opinions (n=3), we think that removing all of the `.iml` files and `.idea` directories in the examples should be fine.  The examples all exist in the top level `modules.xml` now, and there are run configurations for all of them.  We can put them back easily if anyone complains.

Ideally we should have a separate, curated, `flutter/examples` repo that is designed for people to copy things from.  As-is, they can run and modify the `flutter_gallery` and `hello_world` in place, and without the config files, the source is simpler to see what's part of Flutter and what's not, and it's IDE-agnostic. We should also be pointing them to `flutter create` to start new projects anyhow, since starting with a copy of the gallery is not really the best path to success.

So, I've removed all the `.idea` and `.iml`s in the examples, in addition to:
 - Moving all the templates to `packages/flutter_tools/ide_templates`, and updated the code to know that it should also ignore the regular templates directory.
 - Adding identical file detection so that we don't overwrite files that have no differences.
 - Removing the `codeStyleSettings.xml` (since it only points to a profile and doesn't actually set the settings).

gspencergoog committed:
Updating docs to include a pointer to ide-config.

gspencergoog committed:
Removed some unneeded globals.

devoncarew commented on packages/flutter_tools/test/commands/ide_config_test.dart:
```diff
@@ -0,0 +1,311 @@
+      );
+    }, timeout: const Timeout.factor(2.0));
+
+    testUsingContext('removes delted imls with --overwrite', () async {
```
`deleted`?

gspencergoog replied:
Fixed

devoncarew commented on packages/flutter_tools/test/commands/ide_config_test.dart:
```diff
@@ -0,0 +1,311 @@
+      );
+    }, timeout: const Timeout.factor(2.0));
+
+    testUsingContext('removes delted imls with --overwrite, including empty parent dirs', () async {
```
`deleted`?

gspencergoog replied:
Fixed

devoncarew reviewed (approved):
> The examples all exist in the top level modules.xml now, and there are run configurations for all of them.

I don't expect that regular users will run the `flutter ide-config` command though. Making the examples - and specifically flutter_gallery - easy to open is super useful for users, as it's a good app to mine for examples of how to do things (https://github.com/flutter/flutter-intellij/issues/535). `flutter create` doesn't really help them here.

I tested, and - after removing the IntelliJ metadata from flutter_gallery - we're still able to open it from IntelliJ w/o issues (you do need to have already configured a flutter sdk in IntelliJ). So I think it's ok to remove, but we may want to add some back in the future if we find out that it lowers the barrier to getting started for users.

> Made the ide-config command a hidden command.

Thanks, and thanks for for wrangling this metadata issue :)

gspencergoog commented:
I'm glad it can be opened without the metadata: thanks for checking that. Let's see how that works, and if it seems onerous, then we can simply add it back in.

yjbanov commented:
But now one can't use `git clean -xdf` without messing up their workspace?
